domctl: Avoid taking domctl lock for certain ops

When performing multiple migrations in parallel, the domctl lock may
become extremely contended:

* Operations like "xl vcpu-list" were observed to take in excess of 20s
  to execute.
* The "clean" shadow op may pause the domain, restart with a
  continuation and then become blocked on the domctl lock, causing VM
  downtime in excess of 20 seconds.

Fix this contention by not taking the domctl lock for the domctls that
are repeatedly called during migration. This is safe because:

* For XEN_DOMCTL_SHADOW_OP_{CLEAN,PEEK}, the paging lock is held which
  prevents it from operating concurrently on the same domain. There are
  some operations that are called without the paging lock held:

  * hvm_mapped_guest_frames_mark_dirty() - The function itself takes a
    spinlock so is protected from concurrent calls. In any case, it will
    mark all the pages dirty as required.

  * domain_pause() - The toolstack cannot unpause the domain while in
    paging_log_dirty_op() because the toolstack's pause/unpause ops have
    a separate ref count.

  * p2m_flush_hardware_cached_dirty() - This is called elsewhere without
    the domctl lock held so holding it doesn't achieve anything. It
    should be fine as long as it is called at least once.

  * log_dirty.ops->clean() - If the callback is
    hap_clean_dirty_bitmap(), then it will hold the p2m lock while
    modifying the table. If the callback is sh_clean_dirty_bitmap(), it
    will hold the paging lock while modifying the table. In both cases,
    this is OK.

  * domain_unpause() - Same as the earlier domain_pause().

* For XEN_DOMCTL_getpageframeinfo3, it does not change any state so
  multiple concurrent callers is fine (the P2M read lock is taken when
  needed to avoid issues with concurrent changes to the P2M).

Additional ops can be added in future if it is safe to do so.

Signed-off-by: Ross Lagerwall <ross.lagerwall@citrix.com>
diff --git a/xen/arch/arm/domctl.c b/xen/arch/arm/domctl.c
index ad914c915f81..27eb1ededdea 100644
--- a/xen/arch/arm/domctl.c
+++ b/xen/arch/arm/domctl.c
@@ -49,6 +49,11 @@ static int handle_vuart_init(struct domain *d,
     return rc;
 }
 
+bool arch_use_domctl_lock(const struct xen_domctl *op)
+{
+    return true;
+}
+
 long arch_do_domctl(struct xen_domctl *domctl, struct domain *d,
                     XEN_GUEST_HANDLE_PARAM(xen_domctl_t) u_domctl)
 {
diff --git a/xen/arch/x86/domctl.c b/xen/arch/x86/domctl.c
index 5f01111619da..2fda08f39708 100644
--- a/xen/arch/x86/domctl.c
+++ b/xen/arch/x86/domctl.c
@@ -197,6 +197,29 @@ static int do_vmtrace_op(struct domain *d, struct xen_domctl_vmtrace_op *op,
     return rc;
 }
 
+bool arch_use_domctl_lock(const struct xen_domctl *op)
+{
+    switch ( op->cmd )
+    {
+    case XEN_DOMCTL_shadow_op:
+        switch ( op->u.shadow_op.op )
+        {
+        case XEN_DOMCTL_SHADOW_OP_CLEAN:
+        case XEN_DOMCTL_SHADOW_OP_PEEK:
+            return false;
+
+        default:
+            return true;
+        }
+
+    case XEN_DOMCTL_getpageframeinfo3:
+        return false;
+
+    default:
+        return true;
+    }
+}
+
 #define MAX_IOPORTS 0x10000
 
 long arch_do_domctl(
diff --git a/xen/arch/x86/mm/paging.c b/xen/arch/x86/mm/paging.c
index c77f4c1dac52..60c4d928a0e3 100644
--- a/xen/arch/x86/mm/paging.c
+++ b/xen/arch/x86/mm/paging.c
@@ -770,11 +770,14 @@ long do_paging_domctl_cont(
     ret = xsm_domctl(XSM_OTHER, d, op.cmd, 0 /* SSIDref not applicable */);
     if ( !ret )
     {
-        if ( domctl_lock_acquire() )
+        bool use_lock = arch_use_domctl_lock(&op);
+
+        if ( !use_lock || domctl_lock_acquire() )
         {
             ret = paging_domctl(d, &op.u.shadow_op, u_domctl, 1);
 
-            domctl_lock_release();
+            if ( use_lock )
+                domctl_lock_release();
         }
         else
             ret = -ERESTART;
diff --git a/xen/common/domctl.c b/xen/common/domctl.c
index 05abb581a03d..38ab81bbeae5 100644
--- a/xen/common/domctl.c
+++ b/xen/common/domctl.c
@@ -285,6 +285,7 @@ long do_domctl(XEN_GUEST_HANDLE_PARAM(xen_domctl_t) u_domctl)
     bool copyback = false;
     struct xen_domctl curop, *op = &curop;
     struct domain *d;
+    bool use_lock;
 
     if ( copy_from_guest(op, u_domctl, 1) )
         return -EFAULT;
@@ -329,7 +330,8 @@ long do_domctl(XEN_GUEST_HANDLE_PARAM(xen_domctl_t) u_domctl)
     if ( ret )
         goto domctl_out_unlock_domonly;
 
-    if ( !domctl_lock_acquire() )
+    use_lock = arch_use_domctl_lock(op);
+    if ( use_lock && !domctl_lock_acquire() )
     {
         if ( d && d != dom_io )
             rcu_unlock_domain(d);
@@ -881,7 +883,8 @@ long do_domctl(XEN_GUEST_HANDLE_PARAM(xen_domctl_t) u_domctl)
         break;
     }
 
-    domctl_lock_release();
+    if ( use_lock )
+        domctl_lock_release();
 
  domctl_out_unlock_domonly:
     if ( d && d != dom_io )
diff --git a/xen/include/xen/domain.h b/xen/include/xen/domain.h
index ae2b72d5c646..cbc4d994eff6 100644
--- a/xen/include/xen/domain.h
+++ b/xen/include/xen/domain.h
@@ -34,6 +34,7 @@ struct xen_domctl_getdomaininfo;
 void getdomaininfo(struct domain *d, struct xen_domctl_getdomaininfo *info);
 void arch_get_domain_info(const struct domain *d,
                           struct xen_domctl_getdomaininfo *info);
+bool arch_use_domctl_lock(const struct xen_domctl *op);
 
 /* CDF_* constant. Internal flags for domain creation. */
 /* Is this a privileged domain? */
